### PR TITLE
[RDY] Fix problem with coverity/105568 fix.

### PR DESCRIPTION
**Pull request by oni-link**
Fix problem with coverity/105568 fix.

The original fix 3db0a40
does not work for more than one loop iteration, because memory allocated
in the previous iteration could be reused in the current iteration.

Because expand_wildcards() never reads the variables *num_file
and *file before the first assignment to them, the initial
values for these variables can be anything. So instead of
calling expand_shellcmd() with *file = "" we set *file = NULL.
That should help coverity see, that not a array-typed value
is freed.

The commits updating the function comments should help
with reviewing this PR.

### CODE REVIEW

**Comment by elmart** [comment]
- Parameter documentation formatting can be better (parameter names in the same column).
- Maybe better squashing into one commit. Or two (one with function comments improvements and other with the fix).
- `#include <assert.h>` not needed anymore.

None of those is really important, though. I mean, feel free to ignore them if you don't heartfully agree.

The only thing I want to check before giving my ok is that coverity doesn't complain again. Give me half an hour to check.

**Comment by elmart** [comment]
Ok, checked. I can guarantee this doesn't raise the coverity issue again.
Regarding correction of the fix, I'm convinced. I already checked yesterday (when we talked about this), the points where `*file` is used.

So, as soon as Travis build passes, this LGTM.
Whether you do some of the changes suggested above or not.

**Comment by oni-link** [comment]
@elmart, updated with your suggestions.

File: src/nvim/os_unix.c [hunk 1]
```diff
@@ -267,10 +254,27 @@ void mch_exit(int r)
...
+///                      not be freed.
+///                      Set to NULL if FAIL is returned. Otherwise points to
+///                      allocated memory.
+/// @param      flags    is a combination of EW_* flags used in
+///                      expand_wildcards().
+///                      If matching fails but EW_NOTFOUND is set in flags or
+///                      there are no wildcards, the patterns from pat are
+///                      copied into *file.
+///
+/// @returns             Ok for success or FAIL for error.
```
Comment (Pyrohh) [inline src/nvim/os_unix.c]: small nitpick, but shouldn't this be `OK` (capitalized)?

Comment (oni-link) [inline src/nvim/os_unix.c]: Yes, will repush with a fixed version.

**Comment by elmart** [comment]
LGTM.

**Comment by oni-link** [comment]
@Pyrohh, did you finish reviewing?

**Comment by Pyrohh** [comment]
@oni-link I just went over it again and it looks good. Your new function comments are very helpful.

**Comment by Pyrohh** [comment]
Hi, is there anything in particular this requires, another review perhaps?

**Comment by elmart** [comment]
@Pyrohh or @oni-link I think this can be merged.

**Comment by Pyrohh** [comment]
I'll leave merging up to you @elmart or @oni-link. I'm mostly confident in what this PR does, but not *that* confident. After all, I've only been learning C for a few months.

**Comment by justinmk** [comment]
Merged, thanks!

**Pull request closed without merging**